<a href="https://colab.research.google.com/github/chandanakm/DatasetAnalysis/blob/master/algorithmcomparison.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Comparing performance of algorithms**

## 1. KNN algorithm 

In [0]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
dataset = pd.read_csv('https://raw.githubusercontent.com/dipbanik/MyTectraJuly/master/Classification/Social_Network_Ads.csv')
X = dataset.iloc[:,[2,3]].values
y = dataset.iloc[:, 4].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


In [0]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
classifier = KNeighborsClassifier()
classifier.fit(X_train, y_train)
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train, y_train)
y_train_pred = classifier.predict(X_train)
y_test_pred = classifier.predict(X_test)


In [0]:
cm_train = confusion_matrix(y_train, y_train_pred)
cm_train

array([[172,  17],
       [ 10, 101]])

In [0]:
print(classification_report(y_train, y_train_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.95      0.91      0.93       189
negative class       0.86      0.91      0.88       111

      accuracy                           0.91       300
     macro avg       0.90      0.91      0.90       300
  weighted avg       0.91      0.91      0.91       300



In [0]:
cm_test = confusion_matrix(y_test, y_test_pred)
cm_test


array([[64,  4],
       [ 3, 29]])

In [0]:
print(classification_report(y_test, y_test_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.96      0.94      0.95        68
negative class       0.88      0.91      0.89        32

      accuracy                           0.93       100
     macro avg       0.92      0.92      0.92       100
  weighted avg       0.93      0.93      0.93       100



## **KNN Algorithm Test dataset ACCURACY:  0.93**

# Decision tree algorithm

In [0]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)
y_train_pred=classifier.predict(X_train)
y_test_pred = classifier.predict(X_test)
cm_train = confusion_matrix(y_train, y_train_pred)
cm_train

array([[189,   0],
       [  0, 111]])

In [0]:
cm_test = confusion_matrix(y_test, y_test_pred)
cm_test

array([[62,  6],
       [ 3, 29]])

In [0]:
print(classification_report(y_train, y_train_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       1.00      1.00      1.00       189
negative class       1.00      1.00      1.00       111

      accuracy                           1.00       300
     macro avg       1.00      1.00      1.00       300
  weighted avg       1.00      1.00      1.00       300



In [0]:
print(classification_report(y_test, y_test_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.95      0.91      0.93        68
negative class       0.83      0.91      0.87        32

      accuracy                           0.91       100
     macro avg       0.89      0.91      0.90       100
  weighted avg       0.91      0.91      0.91       100



## Decision tree algorithm test dataset ACCURACY: 0.91

# Naive-Bayes **Classifier**

In [0]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix
classifier = GaussianNB()
classifier.fit(X_train, y_train)
y_train_pred = classifier.predict(X_train)
y_test_pred=classifier.predict(X_test)
cm_train = confusion_matrix(y_train, y_train_pred)
cm_test = confusion_matrix(y_test, y_test_pred)

In [0]:
cm_train

array([[174,  15],
       [ 20,  91]])

In [0]:
cm_test

array([[65,  3],
       [ 7, 25]])

In [0]:
print(classification_report(y_train, y_train_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.90      0.92      0.91       189
negative class       0.86      0.82      0.84       111

      accuracy                           0.88       300
     macro avg       0.88      0.87      0.87       300
  weighted avg       0.88      0.88      0.88       300



In [0]:
print(classification_report(y_test, y_test_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.90      0.96      0.93        68
negative class       0.89      0.78      0.83        32

      accuracy                           0.90       100
     macro avg       0.90      0.87      0.88       100
  weighted avg       0.90      0.90      0.90       100



### Naive-Bayes Algortihm test dataset ACCURACY: 0.**90**

# **SVM Algorithm**

In [0]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)
y_train_pred = classifier.predict(X_train)
y_test_pred = classifier.predict(X_test)
cm_train = confusion_matrix(y_train, y_train_pred)
cm_test = confusion_matrix(y_test, y_test_pred)


In [0]:
cm_train

array([[173,  16],
       [ 37,  74]])

In [0]:
cm_test

array([[66,  2],
       [ 8, 24]])

In [0]:
print(classification_report(y_train, y_train_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.82      0.92      0.87       189
negative class       0.82      0.67      0.74       111

      accuracy                           0.82       300
     macro avg       0.82      0.79      0.80       300
  weighted avg       0.82      0.82      0.82       300



In [0]:
print(classification_report(y_test, y_test_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.89      0.97      0.93        68
negative class       0.92      0.75      0.83        32

      accuracy                           0.90       100
     macro avg       0.91      0.86      0.88       100
  weighted avg       0.90      0.90      0.90       100



## SVM test dataset ACCURACY : 0.**90**

# Random forest Algorithm 

In [0]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)
y_train_pred = classifier.predict(X_train)
y_test_pred = classifier.predict(X_test)
cm_train = confusion_matrix(y_train, y_train_pred)
cm_test = confusion_matrix(y_test, y_test_pred)


In [0]:
cm_train


array([[187,   2],
       [  4, 107]])

In [0]:
cm_test

array([[63,  5],
       [ 4, 28]])

In [0]:
print(classification_report(y_train, y_train_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.98      0.99      0.98       189
negative class       0.98      0.96      0.97       111

      accuracy                           0.98       300
     macro avg       0.98      0.98      0.98       300
  weighted avg       0.98      0.98      0.98       300



In [0]:
print(classification_report(y_test, y_test_pred, target_names=['positive class','negative class']))

                precision    recall  f1-score   support

positive class       0.94      0.93      0.93        68
negative class       0.85      0.88      0.86        32

      accuracy                           0.91       100
     macro avg       0.89      0.90      0.90       100
  weighted avg       0.91      0.91      0.91       100



## Random forest test dataset ACCURACY: 0.**91**

1. KNN          - 0.93
2. Decision Tree- 0.91
3. Naive Bayes  - 0.90
4. SVM          - 0.90
5. Random Forest- 0.91


Comparitively KNN has highest Accuracy for test dataset.